File: frontend/src/utils/passwordPolicy.ts
import { z } from 'zod'

export const PASSWORD_MIN_LENGTH = 10
export const PASSWORD_POLICY_HELP =
  'Minimo 10 caracteres, com maiuscula, minuscula, numero e caractere especial.'

const UPPERCASE_RE = /[A-Z]/
const LOWERCASE_RE = /[a-z]/
const NUMBER_RE = /\d/
const SPECIAL_RE = /[!@#$%^&*(),.?":{}|<>]/

export function getPasswordPolicyError(password: string): string | null {
  if (password.length < PASSWORD_MIN_LENGTH) {
    return `Minimo ${PASSWORD_MIN_LENGTH} caracteres.`
  }
  if (!UPPERCASE_RE.test(password)) {
    return 'Inclua ao menos uma letra maiuscula.'
  }
  if (!LOWERCASE_RE.test(password)) {
    return 'Inclua ao menos uma letra minuscula.'
  }
  if (!NUMBER_RE.test(password)) {
    return 'Inclua ao menos um numero.'
  }
  if (!SPECIAL_RE.test(password)) {
    return 'Inclua ao menos um caractere especial.'
  }
  return null
}

export function isStrongPassword(password: string): boolean {
  return getPasswordPolicyError(password) === null
}

export const strongPasswordSchema = z.string().superRefine((password, ctx) => {
  const message = getPasswordPolicyError(password)
  if (message) {
    ctx.addIssue({ code: z.ZodIssueCode.custom, message })
  }
})

File: backend/app/schemas/password_policy.py
import re

PASSWORD_MIN_LENGTH = 10
PASSWORD_SPECIAL_CHARS = r"!@#$%^&*(),.?\":{}|<>"


def validate_strong_password(value: str) -> str:
    if len(value) < PASSWORD_MIN_LENGTH:
        raise ValueError(f"Senha deve ter no minimo {PASSWORD_MIN_LENGTH} caracteres")
    if not re.search(r"[A-Z]", value):
        raise ValueError("Senha deve conter pelo menos uma letra maiuscula")
    if not re.search(r"[a-z]", value):
        raise ValueError("Senha deve conter pelo menos uma letra minuscula")
    if not re.search(r"\d", value):
        raise ValueError("Senha deve conter pelo menos um numero")
    if not re.search(f"[{re.escape(PASSWORD_SPECIAL_CHARS)}]", value):
        raise ValueError(
            "Senha deve conter pelo menos um caractere especial "
            f"({PASSWORD_SPECIAL_CHARS})"
        )
    return value
